### PR TITLE
[VL] Pass 75% of off-heap memory size as Velox memory tracker's limit

### PR DESCRIPTION
**Pull request by zhztheplayer**

### CODE REVIEW

File: cpp/velox/compute/WholeStageResultIterator.cc [hunk 1]
```diff
@@ -14,7 +14,7 @@ namespace gluten {
 
 namespace {
 const std::string kSparkBatchSizeKey = "spark.sql.execution.arrow.maxRecordsPerBatch";
-const std::string kSparkOffHeapSizeKey = "spark.memory.offHeap.size";
+const std::string kSparkOffHeapMemoryPerTask = "spark.gluten.memory.offHeap.size.in.bytes.pertask";
```
Comment (zhztheplayer) [inline cpp/velox/compute/WholeStageResultIterator.cc:17]: @rui-mo 

I see you have set 0.75 * executor-offheap-size as `kMaxPartialAggregationMemory`.

https://github.com/oap-project/gluten/blob/15faa98cda0a1a2fe8f7c2b3e4c73ce159b791f7/cpp/velox/compute/WholeStageResultIterator.cc#L182-L190

In this PR I am changing `kSparkOffHeapSizeKey` to `kSparkOffHeapMemoryPerTask` (per executor -> per task). Should we update the factor 0.75 to another number? Was it intended to use executor's off-heap memory size for partial agg?

Comment (rui-mo) [reply cpp/velox/compute/WholeStageResultIterator.cc:17]: This setting was used to enlarge the default value and to achieve more flexible flush, see https://github.com/oap-project/velox/pull/98. Looks an adjustment is needed with this PRs' change but I'm not sure about the value. Do you have any suggestion? @FelixYBW

**Comment by FelixYBW** [comment]
what's the goal of this PR? what's the issue to solve?

**Comment by zhztheplayer** [comment]
> what's the goal of this PR? what's the issue to solve?

To trigger Velox's spill. Velox requires for a limit to be set otherwise it doesn't spill data.

I'm still tweaking on the factor, 1/2 is just a magic number I came up with, finally it could be 0.75 , 0.8, 0.85 or could be configurable, not sure at this time.

**Comment by FelixYBW** [comment]
> 

Let's at least make ti configurable for now. Let's use 0.75 for now.

File: cpp/velox/compute/WholeStageResultIterator.cc (hunk 1 above)
Comment (zhztheplayer) [reply cpp/velox/compute/WholeStageResultIterator.cc]: I restored the old code of this part and left a comment there. Let's propose fix later on.

**Comment by zhztheplayer** [comment]
@rui-mo

File: ep/build-velox/src/build_velox.sh [hunk 2]
```diff
@@ -10,7 +10,7 @@ BUILD_PROTOBUF=ON
 BUILD_TYPE=release
 VELOX_HOME=
 #for ep cache
-VELOX_REPO=https://github.com/oap-project/velox.git
+VELOX_REPO=https://github.com/zhztheplayer/velox.git
```
Comment (rui-mo) [inline ep/build-velox/src/build_velox.sh:13]: Do you have PR for the Velox change?

**Review by rui-mo — APPROVED** [review]
LGTM. Just some tiny issues.

File: cpp/velox/compute/VeloxBackend.h [hunk 3]
```diff
@@ -50,7 +52,7 @@ class VeloxBackend final : public Backend {
       std::shared_ptr<ColumnarBatch> cb) override;
 
   std::shared_ptr<SplitterBase> makeSplitter(
-      const std::string& partitioning_name,
+      const std::string& partitioning_na00me,
```
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply cpp/velox/compute/VeloxBackend.h]: oops, thanks for catching

File: ep/build-velox/src/build_velox.sh (hunk 2 above)
Comment (zhztheplayer) [reply ep/build-velox/src/build_velox.sh:13]: Not yes, I'll file one once this get approved.

File: cpp/velox/compute/VeloxBackend.cc [hunk 4]
```diff
@@ -28,12 +28,33 @@
...
+  auto got = confMap_.find(kSparkOffHeapMemory);
+  if (got == confMap_.end()) {
+    // not found
+    maxMemory = facebook::velox::memory::kMaxMemory;
+  } else {
+    maxMemory = (long)(0.75 * std::stol(got->second));
+  }
+  try {
+    // 1/2 of offheap size.
+    mem_usage_tracker_ = velox::memory::MemoryUsageTracker::create(maxMemory);
```
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply cpp/velox/compute/VeloxBackend.cc:52]: Do you mean `mem_usage_tracker_`?

[the remark this answers is not in the document]
Comment (zhztheplayer) [reply cpp/velox/compute/VeloxBackend.cc]: changed to memUsageTracker_

File: ep/build-velox/src/build_velox.sh (hunk 2 above)
Comment (zhztheplayer) [reply ep/build-velox/src/build_velox.sh:13]: https://github.com/oap-project/velox/pull/154

I've open the PR. Would you want to review that one too?

Comment (rui-mo) [reply ep/build-velox/src/build_velox.sh:13]: Got it. Thanks!

Comment (rui-mo) [reply ep/build-velox/src/build_velox.sh:13]: Merged this PR.

**Review by rui-mo — DISMISSED** [review]

**Review by rui-mo — APPROVED** [review]

**Review by rui-mo — APPROVED** [review]

**Pull request merged**